memas/corpus/corpus_searching.py

```diff
@@ -1,17 +1,55 @@
 # from search_redirect import SearchSettings
 from uuid import UUID
 from functools import reduce
-from memas.interface.corpus import Corpus, CorpusFactory
+from memas.interface.corpus import Corpus, CorpusFactory, CorpusType
 from memas.interface.corpus import Citation
+from collections import defaultdict 
 from memas.interface.storage_driver import DocumentEntity
 from memas.interface.exceptions import SentenceLengthOverflowException
 
 
-def corpora_search(corpus_ids: list[UUID], clue: str) -> list[tuple[float, str, Citation]]:
+def mult_corpus_search(corpus_sets : dict[Corpus], clue, ctx, result_limit) -> list[tuple[float, str, Citation]]:
+    results = defaultdict(list)
+
+    # Direct each multicorpus search to the right algorithm
+    for corpusType, corpora_list in corpus_sets.items() :
+        # Default basic corpus handling
+        if corpusType == CorpusType.KNOWLEDGE or corpusType == CorpusType.CONVERSATION :
+            corpus_type_results = basic_corpora_search(corpora_list, clue, ctx)
+            results["BASIC_SCORING"].extend(corpus_type_results)
+
+
+    sorted_results_matrix = []
+    # Sort results with compareable scoring schemes
+    for scored_results in results.values() :
+        # Sort by descending scoring so best results come first
+        sorted_scored_results = sorted(scored_results, key=lambda x: x[0], reverse=True)
+        sorted_results_matrix.append(sorted_scored_results)
+
+    # To combine results for corpora that don't have compareable scoring take equal sized subsets of each Corpus type
+    # TODO : Consider changing this at some point in the future to have better searching of corpus sets with non-comparable scoring
+    combined_results = []
+    for j in range(max([len(x) for x in sorted_results_matrix])) :
+        for i in range(len(sorted_results_matrix)) :
+            if j >= len(sorted_results_matrix[i]) or len(combined_results) >= result_limit:
+                break
+            combined_results.append(sorted_results_matrix[i][j])
+        if len(combined_results) >= result_limit:
+            break
```

Comment:
This segment for combining the results confuse me. What's the purpose of these two loops, is it to extract an equal number of results from each corpus type equally? In which case wouldn't just extracting the top `result_limit/n`, where `n` is the number of corpus types?

Reply:
The two loops are for extracting result_limit results, preferably equally for each different (non-comparable) way of corpus scoring.  Your suggestion was my initial plan for how to do it, but it gets more complicated when you can't guarantee there are result_limit/n results to fetch from each corpus. The loops are one way of dealing with that while also ordering the results.

Reply:
Gotcha, sounds good